assign-2/src/data/ArrayListStudentDao.java

```diff
@@ -0,0 +1,51 @@
+package data;
+
+import data.StudentDao;
+import logic.Student;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ArrayListStudentDao implements StudentDao {
+    private List<Student> students = new ArrayList<>();
+
+    @Override
+    public void add(Student student) {
+        students.add(student);
+    }
+
+    @Override
+    public void update(Student student) {
+        Student s = getById(student.getId());
+        if (s != null) {
+            s.setFirstName(student.getFirstName());
+            s.setLastName(student.getLastName());
+        }
+    }
+
+    @Override
+    public void delete(int id) {
+        Student s = getById(id);
+        if (s != null) {
+            students.remove(s);
+        } else {
+            throw new NoSuchStudentException(id);
+        }
+    }
+
+
+    @Override
+    public Student getById(int id) {
+        for (Student student : students) {
+            if (student.getId() == id) {
+                return student;
+            }
+        }
+        return null;
```

Comment:
Instead of returning `null` you could also use your `NoSuchStudentException`. And wherever you are calling the `getById` method, you can catch the exception and implement proper error handling. You do not have to correct/implement this right now, since the next teaching block is about exception handling. But maybe you can think already about how to use exception handling instead of returning `null`.